python-checks/src/main/java/org/sonar/python/checks/SklearnPipelineParameterAreCorrectCheck.java

```diff
@@ -0,0 +1,182 @@
+import org.sonar.plugins.python.api.tree.ExpressionList;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.RegularArgument;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.checks.utils.Expressions;
+import org.sonar.python.tree.ListLiteralImpl;
+import org.sonar.python.tree.TreeUtils;
+import org.sonar.python.tree.TupleImpl;
+
+@Rule(key = "S6972")
+public class SklearnPipelineParameterAreCorrectCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Provide valid parameters to the estimator.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CALL_EXPR, SklearnPipelineParameterAreCorrectCheck::checkCallExpression);
+  }
+
+  private static void checkCallExpression(SubscriptionContext subscriptionContext) {
+    CallExpression callExpression = (CallExpression) subscriptionContext.syntaxNode();
+
+    var isRelevantCall = Optional.of(callExpression).map(CallExpression::calleeSymbol).map(Symbol::fullyQualifiedName)
+      .filter("sklearn.pipeline.Pipeline.set_params"::equals);
+
+    if (isRelevantCall.isEmpty()) {
+      return;
+    }
+
+    var parsedFunction = getStepAndParametersFromArguments(callExpression);
+    var pipelineName = Optional.of(callExpression).map(CallExpression::callee).flatMap(TreeUtils.toOptionalInstanceOfMapper(QualifiedExpression.class))
+      .map(QualifiedExpression::qualifier).flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class));
+
+    pipelineName.flatMap(Expressions::singleAssignedNonNameValue)
+      .ifPresent(c1 -> pipelineName.ifPresent(pipelineName1 -> {
+        var parsedPipeline = parsePipeline(((CallExpression) c1));
+        findProblems(parsedFunction, parsedPipeline, subscriptionContext);
+      }));
+  }
+
+  private static void findProblems(Map<String, Set<StringAndTree>> setParameters, Map<String, ClassSymbol> pipelineDefinition, SubscriptionContext subscriptionContext) {
+    for (var entry : setParameters.entrySet()) {
+      var step = entry.getKey();
+      var stringAndTree = entry.getValue();
+      var parameters = stringAndTree.stream().map(StringAndTree::string).collect(Collectors.toSet());
+
+      var classifier = pipelineDefinition.get(step);
+      if (classifier == null) {
+        // Maybe quickfix ?
+        continue;
+      }
+      var possibleParameters = getInitFunctionSymbol(classifier).map(FunctionSymbol::parameters).orElse(List.of());
+
+      parameters.forEach(parameter -> {
+        if (possibleParameters.stream().noneMatch(symbol -> Objects.equals(symbol.name(), parameter))) {
+          stringAndTree.stream().filter(stringAndTree1 -> stringAndTree1.string().equals(parameter)).findFirst()
+            .ifPresent(location -> subscriptionContext.addIssue(location.tree, MESSAGE));
```

Comment:
This is not straightforward to understand as well. We could also maybe create small functions just to name things.